Implement HTML Report Generation

StarKhan6368 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Implement HTML Report Generation

<!-- Why are these changes necessary? -->

**Why**:

This pull request adds HTML report generation to improve the reviewing of test results. HTML interface provides better visibility and understanding of test outcomes compared to console output.

#### Feature

1. Generates comprehensive HTML reports of test outcomes, including coverage metrics.
2. Provides functionality to inspect and analyze individual request, response and duration details.
3. Includes a comprehensive summary of the tests, incorporating total duration, failures, errors, and more. 
4. Printable Test Coverage Report, customizability and more.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/znsio/specmatic-documentation
- [ ] Tests
- [ ] Sonar Quality Gate

<!-- feel free to add additional comments -->

StarKhan6368 committed:
Add `HTML` report generation, test data recorder.

- Heavy modifications to `core`.
- Add html assets, icons, js, css, etc.
- Implement HTML Report generator.
- New Fields to `TestResultRecord`,
  `HttpLogMessage` and `OpenApiCoverageConsoleRow`
- Modifications to `HttpClient` Executor.
- DataRecorder to record HttpLogMessage(s).

StarKhan6368 committed:
Fix `Coverage` percentage and response details.

- Rename `remark` to `result` in details.
- Add New Field to OpenApiCoverageConsoleRow.

StarKhan6368 committed:
Fix Overflow problem with Scenario names.

- Limit max-width on Scenario Item Element.

joelrosario committed:
Merge branch 'main' into html-report

joelrosario committed:
Add a null check in case a method in the spec was not executed as a test

joelrosario committed:
Added a script to build a docker image locally

joelrosario committed:
Renamed the docker image build script

StarKhan6368 committed:
Enable Word Wrap for long path names.

StarKhan6368 committed:
Increase `path` column size to be 2 columns wide.

StarKhan6368 committed:
Add new TestResult and Remark `Invalid`.

- Any test under any method under any endpoint
  without a path param should be marked invalid.
- Fix summary container when printing html report.
- Fix testResults with responseStatus `0` code.

StarKhan6368 committed:
Fix isEndpointInvalid function logic.

joelrosario committed:
Implemented SPECAMTIC_GENERATIVE_STUB

joelrosario committed:
Merge branch 'html-report' of github.com:znsio/specmatic into html-report

joelrosario committed:
Updated the docker build script to assemble specmatic

joelrosario committed:
Dump the error in the message field if it exists

StarKhan6368 committed:
Change the way invalid tests are labelled.

- Add `isValid` field to TestResultRecord
- Remove `Invalid` TestResult
- Enable wrap for request / response in html.
- Add success count to HTML report.

StarKhan6368 committed:
Merge remote-tracking branch 'origin/html-report' into html-report

StarKhan6368 committed:
rename `isValid` to `valid` in ScenarioData.

StarKhan6368 committed:
change html heading to `CTK Results`

joelrosario committed:
Revert "Renamed the docker image build script"

This reverts commit f5a95b2d6d37dfd7917432322e0641fe63186aaf.

joelrosario committed:
Revert "Added a script to build a docker image locally"

This reverts commit e57f1a43219bf9348472fd0458d37657e122f46a.

joelrosario committed:
Revert "Implemented SPECAMTIC_GENERATIVE_STUB"

This reverts commit 1f63c82f244e4f136e31b231485fc7ae03c89d40.

joelrosario committed:
Revert "Updated the docker build script to assemble specmatic"

This reverts commit c8f21493081158f5a7aa76200c751b82a5b82203.

joelrosario committed:
Revert "Dump the error in the message field if it exists"

This reverts commit 0975df8a3257b758b2d952a3d3020fb7a5a841da.

joelrosario committed:
Merge branch 'main' into html-report

StarKhan6368 committed:
Merge branch 'refs/heads/main' into html-report

# Conflicts:
#	core/src/main/kotlin/io/specmatic/test/TestResultRecord.kt
#	junit5-support/src/main/kotlin/io/specmatic/test/reports/coverage/OpenApiCoverageReportInput.kt
#	junit5-support/src/main/kotlin/io/specmatic/test/reports/coverage/console/Remarks.kt

StarKhan6368 committed:
Add html config attributes to SpecmaticConfig.

- Fix failing tests for HttpLogMessage.
- Fix failing tests for HttpClient.
- Remove SPECMATIC_REPORT_DIRECTORY Flag.
- Add `title`, `heading` and `outputDirectory`
  attributes for html in SpecmaticConfig.

StarKhan6368 committed:
refactor ConsoleRow, remove html hardcoding.

- Use boolean values to hide / show path, method
  and coverage for console report.
- flexible title, heading, outDir for html report.
- Refactor tests according to ConsoleRow changes.
- minor refactorings.

StarKhan6368 committed:
remove capitalization from `paths` in html table.

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/test/DataRecorder.kt:
```diff
@@ -0,0 +1,21 @@
+
+import io.specmatic.core.log.HttpLogMessage
+
+object DataRecorder {
```
Please rename to `TestInteractionsLog`

StarKhan6368 replied:
Thanks, renamed it to `TestInteractionsLog`

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/test/DataRecorder.kt:
```diff
@@ -0,0 +1,21 @@
+object DataRecorder {
+
+    val testHttpLogMessages = mutableListOf<HttpLogMessage>()
+    val stubHttpLongMessages = mutableListOf<HttpLogMessage>()
```
I guess you meant to name this `stubHttpLogMessages`?

StarKhan6368 replied:
thanks, fixed the typo

harikrishnan83 commented on core/src/main/kotlin/io/specmatic/test/DataRecorder.kt:
```diff
@@ -0,0 +1,21 @@
+    val stubHttpLongMessages = mutableListOf<HttpLogMessage>()
+
+    fun addHttpLog(httpLogMessage: HttpLogMessage) {
+        if(httpLogMessage.scenario != null) {
```
This `if` block and null check (along with redundant else) we need to remove / restructure it

StarKhan6368 replied:
Thanks, simplified the logic and made it more readable.

StarKhan6368 committed:
Rename `DataRecorder` to `TestInteractionsLog`.

- simplify `addHttpLog` function, readable.
- add `isTestLog` function in `HttpLogMessage` to
   check if a log is from test or stub execution.

harikrishnan83 reviewed (approved):
Looks good to me. Thanks.
